### PR TITLE
Fragment for grammar generation with support for semantic predicates

### PR DESCRIPTION
**Pull request by rmitin**
The following extension is supported:

A datatype grammar rule containing only one ID terminal rule can be annotated with @KeywordRule annotation provided a list of words so only these words can be accepted by this rule.

Example:

```
   /**
     * @KeywordRule(visible, invisible)
     * /
   VisibleKind returns VisibleKind:
     ID
   ;

```
 
The above rule will accept only 'visible' and 'invisible' identifiers. This rule we called a keyword rule because it is intended to replace usages of keywords which shall not be reserved words in the language. Reserved words are words that are not allowed to be used in identifiers.

The above example can therefore replace the following enumeration:

```
   enum VisibleKind :
       VISIBLE   = "visible"
     | INVISIBLE = "invisible"
   ;

```

The generator will insert validating semantic predicates in the rule and gated semantic predicates whenever needed in the grammar. Syntax error messages will be adjusted accordingly.

Also the fragment support more generic 

```
   /**
     * @SemanticPredicate(predicateName)
     * /
   Some Rule returns SomeEObject:
     anything
   ;
```

to insert validating and gated predicates.

### CODE REVIEW

**Comment by rmitin** [comment]
Two things are still missing
 - content assist is not fully functional yet
 - binding in runtime module should be done by workflow

File: com.avaloq.tools.ddk.xtext.generator/src/com/avaloq/tools/ddk/xtext/generator/parser/common/PredicatesNaming.xtend [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+package com.avaloq.tools.ddk.xtext.generator.parser.common
+
+import com.google.inject.Inject
+import org.eclipse.xtext.Grammar
+import org.eclipse.xtext.GrammarUtil
+import org.eclipse.xtext.generator.Naming
+
+class PredicatesNaming {
+
+  @Inject extension Naming naming
```
[the remark this answers is not in the document]
Comment (rmitin) [reply com.avaloq.tools.ddk.xtext.generator/src/com/avaloq/tools/ddk/xtext/generator/parser/common/PredicatesNaming.xtend]: Right. Done.

File: com.avaloq.tools.ddk.xtext.generator/src/com/avaloq/tools/ddk/xtext/generator/parser/antlr/AnnotationAwareXtextAntlrGeneratorFragment2.xtend [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    super.checkGrammar();
+    getGrammar().annotateGrammar
+  }
+
+  protected override doGenerate() {
+      super.doGenerate()
+      generateAbstractSemanticPredicate().writeTo(projectConfig.runtime.srcGen)
+  }
+
+  protected override generateProductionGrammar() {
```
Comment (andrewL-avlq) [inline com.avaloq.tools.ddk.xtext.generator/src/com/avaloq/tools/ddk/xtext/generator/parser/antlr/AnnotationAwareXtextAntlrGeneratorFragment2.xtend]: Not sure I understand why this method has been overridden when the implementation is the same, and presumably all the correct classes will be injected into the super class.

Comment (rmitin) [reply com.avaloq.tools.ddk.xtext.generator/src/com/avaloq/tools/ddk/xtext/generator/parser/antlr/AnnotationAwareXtextAntlrGeneratorFragment2.xtend]: productionGenerator is the local field here and is injected again as the type is different. AnnotationAwareAntlrGrammarGenerator cannot be made a subtype of the type used in the parent class (AntlrGrammarGenerator) as we need to do super... calls to the super.super type.

Comment (andrewL-avlq) [reply com.avaloq.tools.ddk.xtext.generator/src/com/avaloq/tools/ddk/xtext/generator/parser/antlr/AnnotationAwareXtextAntlrGeneratorFragment2.xtend]: Ah OK. Makes sense.

**Pull request merged**